File: 섬연결하기.py
# min heap

import heapq

def solution(n, costs):
    answer = 0
    
    # dijkstra
    graph = [[] for _ in range(n)]

    visited = [False for _ in range(n)]
    
    
    for cost in costs:
        # undirected graph
        # 양쪽 연결
        graph[cost[0]].append((cost[1], cost[2]))
        graph[cost[1]].append((cost[0], cost[2]))
        
    
        
    # 방문 가능한 모든 간선 리스트
    # 간선이 짧은 순으로 방문
    # 이미 방문한 노드는 방문하지 않음
    hq = [(0, 0)] # (간선 가중치, 도착지노드)
    
    cnt = 0
    while hq:
        
        w, goal = heapq.heappop(hq)
        
        if visited[goal]:
            continue
            
            
        visited[goal] = True
        
        answer += w
        cnt += 1
        
        if cnt == n:
            break
        
        for node, weight in graph[goal]:
            
            if not visited[node]:
                heapq.heappush(hq, (weight, node))
        
        
    
    
    return answer

File: 섬연결하기.java
import java.util.*;


record Node(int w, int n){}


class Solution {
    public int solution(int n, int[][] costs) {
        int answer = 0;
        
        
        List<List<Node>> graph = new ArrayList<>();
        
        // initialize
        for(int i=0; i<n; i++)
            graph.add(new ArrayList<>());
        
        
        for(int[] cost: costs){
            graph.get(cost[0]).add(new Node(cost[2], cost[1])); // weight, node
            graph.get(cost[1]).add(new Node(cost[2], cost[0])); // weight, node
        }
        
        
        // dijkstra
        // min heap (weight 기준 ascending sort)
        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> a.w() - b.w());
        pq.offer(new Node(0, 0));
        
        
        int cnt = 0;
        
        
        boolean visited[] = new boolean[n];
        while(cnt < n){
            
            Node p = pq.poll();
            int w = p.w();
            int node = p.n();
            
            if(visited[node])
                continue;
            
            
            answer += w;
            cnt++;
            visited[node] = true;
            
            for(Node child: graph.get(node)){
                int weight = child.w();
                int goal = child.n();
                
                if(!visited[goal])
                    pq.offer(new Node(weight, goal));
            }
            
        }
        
        
        
        return answer;
    }
}
